File: shared/src/main/java/com/greencross/lims/data/Work.java
package com.greencross.lims.data;

import jsinterop.annotations.*;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;

@JsType(isNative=true, namespace= JsPackage.GLOBAL, name="Object")
@Setter(onMethod_= {@JsOverlay, @JsIgnore})
@Getter(onMethod_= {@JsOverlay, @JsIgnore})
@Accessors(fluent=true)
public final class Work {
    private String worklist;
    @JsProperty(name="index")
    private Double index;
    private String samples;
    private String services;
    private String mrns;
    @JsProperty(name = "patient_name")
    private String patientName;
    private String gid;
    private String json;
    @JsProperty(name = "last_modify_by")
    private User lastModifyBy;
    @JsProperty(name = "created_by")
    private User createdBy;

    @JsOverlay
    @JsIgnore
    public Integer index(){
        if(index == null) return null;
        return index.intValue();
    }

    private static class User{
        private String id;
        private String name;
    }
}

File: worklist-api/src/main/kotlin/com/greencross/lims/projection/Work.kt
package com.greencross.lims.projection

import com.greencross.lims.entity.User
import io.r2dbc.postgresql.codec.Json
import java.time.LocalDateTime

data class Work(
    val worklist:       String,
    val index:          Int,
    val samples:        String,
    val services:       String,
    val patientName:    String?,
    val mrns:           String?,
    val gid:            String,
    val json:           String?,
    val createBy:       User,
    val createAt:       LocalDateTime,
    val lastModifyBy:   User?,
    val lastModifyAt:   LocalDateTime?
    ) {
    companion object{
        data class WorkBuilder(
            val worklist:       String,
            val index:          Int,
            val samples:        String,
            val services:       String,
            val patientName:    String?,
            val mrns:           String?,
            val gid:            String,
            var json:           Json?,
            val createAt:       LocalDateTime,
            val createById:     String,
            val createBy:       String,
            val lastModifyAt:   LocalDateTime?,
            val lastModifyById: String?,
            val lastModifyBy:   String?
        ){
            fun build() = Work(worklist, index, samples, services, patientName, mrns, gid, json?.asString(), User(createById, createBy), createAt, User(lastModifyById, lastModifyBy), lastModifyAt)
        }
    }
}